### PR TITLE
feat: implement AO rewards

### PR DESCRIPTION
**Pull request by philipstanislaus**
Spec: https://centrifuge.hackmd.io/sJqR0FQZRxakQcSH-8ga-Q
Deployments:
* Kovan, hosted graph node: https://thegraph.com/explorer/subgraph/centrifuge/tinlake-kovan-staging
* Mainnet, hosted graph node: https://thegraph.com/explorer/subgraph/philipstanislaus/tinlake
* Mainnet, own graph node: http://thegraph.centrifuge.io/subgraphs/name/centrifuge/tinlake-ao-rewards/graphql

### CODE REVIEW

File: schema.graphql [hunk 1]
```diff
@@ -214,6 +222,46 @@ type RewardBalance @entity {
   nonZeroBalanceSince: BigInt
 }
 
+"""
```
Comment (offerijns) [inline schema.graphql:225]: Thanks for the extensive comments here 🙌

**Review by offerijns — APPROVED** [review]
I don't have much to say, this looks great @philipstanislaus ! I didn't get a chance yet to test the subgraph data, somehow it isn't loading for me. I'll try that again later.

File: src/config.ts [hunk 2]
```diff
@@ -11,6 +11,7 @@ export const blockTimeSeconds = 15
 export let fixed27 = BigInt.fromI32(10).pow(27)
 
 export const rewardsCeiling = '5600000000000000000000000000000000'
+export const aoRewardsCeiling = '5600000000000000000000000000000000' // TODO discuss height
```
Comment (jpangelle) [inline src/config.ts:14]: curious what `ao` describes

File: src/domain/AOReward.ts [hunk 3]
```diff
@@ -0,0 +1,78 @@
+import { BigInt, BigDecimal, log } from '@graphprotocol/graph-ts'
+import { Pool, PoolAddresses, RewardDayTotal, RewardLink, AORewardBalance } from '../../generated/schema'
+import { aoRewardsCeiling, secondsInDay } from '../config'
+import { loadOrCreateRewardDayTotal } from './Reward'
+
+export function loadOrCreateAORewardBalance(address: string): AORewardBalance {
+  let rb = AORewardBalance.load(address)
```
Comment (jpangelle) [inline src/domain/AOReward.ts:7]: Would rather more descriptive variable names 😄

File: src/domain/AOReward.ts [hunk 4]
```diff
@@ -0,0 +1,78 @@
...
+  log.debug('calculateAORewards: running for pool {}, on {}', [pool.id.toString(), date.toString()])
+
+  let systemRewards = loadOrCreateRewardDayTotal(date)
+  systemRewards = setAORewardRate(systemRewards)
+
+  let tokenAddresses = PoolAddresses.load(pool.id)
+
+  let reward = loadOrCreateAORewardBalance(tokenAddresses.id)
+
+  let r = pool.totalDebt.toBigDecimal().times(systemRewards.aoRewardRate)
```
Comment (jpangelle) [inline src/domain/AOReward.ts:28]: same here for variable name

**Review by jpangelle — COMMENTED** [review]
I don't have much context on this work so I left some nitpicky comments about variable names for your last day 😅 😆

File: src/config.ts (hunk 2 above)
Comment (philipstanislaus) [reply src/config.ts:14]: Added a definition to schema.graphql

File: src/domain/AOReward.ts (hunk 3 above)
Comment (philipstanislaus) [reply src/domain/AOReward.ts:7]: Yeah, this is an artifact from an earlier version – therefore not changing it here now :D

**Pull request merged**